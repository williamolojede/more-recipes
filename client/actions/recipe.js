import { batchActions } from 'redux-batched-actions';

import instance from '../config/axios';

import { setFetching, unsetFetching } from './fetching';
import { showNotification } from './notification';

import {
  RECIEVE_TOP_RATED_RECIPE,
  RECIEVE_SINGLE_RECIPE,
  RECIEVE_NEW_RECIPE,
  RECIEVE_UPDATED_RECIPE
} from './types';

const recieveTopRecipes = (recipes, metaData) => ({
  type: RECIEVE_TOP_RATED_RECIPE,
  recipes,
  metaData
});

const recieveSingleRecipe = (type, recipe) => ({
  type,
  recipe
});


const recieveUpdatedRecipe = (recipe, index) => ({
  type: RECIEVE_UPDATED_RECIPE,
  recipe,
  index
});

export const fetchSingleRecipes = id => (dispatch) => {
  dispatch(setFetching());
  return instance.get(`/recipes/${id}`)
    .then((res) => {
      dispatch(
        batchActions([
          recieveSingleRecipe(RECIEVE_SINGLE_RECIPE, res.data.recipe),
          unsetFetching()
        ])
      );
    })
    .catch((err) => {
      console.log(err.message || err.response.data.message);
      dispatch(unsetFetching());
    });
};

export const fetchTopRecipes = (page, limit) => (dispatch) => {
  dispatch(setFetching());
  return instance.get(`/recipes?sort=upvotes&order=descending&page=${page}&limit=${limit}`)
    .then((res) => {
      dispatch(
        batchActions([
          recieveTopRecipes(res.data.recipes, res.data.pagination),
          unsetFetching()
        ])
      );
    })
    .catch((err) => {
      console.log(err.message || err.response.data.message);
      dispatch(unsetFetching());
    });
};

export const vote = (dir, id) => dispatch => instance.post(`/recipes/${id}/vote-${dir}`)
  .then((res) => {
    dispatch(batchActions([
      recieveSingleRecipe(RECIEVE_SINGLE_RECIPE, res.data.recipe),
      showNotification(res.data.message)
    ]));
  })
  .catch((err) => {
    dispatch(showNotification(err.response.data.message));
  });


export const favorite = id => dispatch => instance.post(`/recipes/${id}/favorite`)
  .then((res) => {
    dispatch(batchActions([
      recieveSingleRecipe(RECIEVE_SINGLE_RECIPE, res.data.recipe),
      showNotification(res.data.message)
    ]));
  })
  .catch((err) => {
    dispatch(showNotification(err.response.data.message));
  });

export const addRecipe = recipe => dispatch =>
  instance.post('/recipes', { recipe })
    .then((res) => {
      dispatch(batchActions([
        recieveSingleRecipe(RECIEVE_NEW_RECIPE, res.data.recipe),
        showNotification(res.data.message),
      ]));
    })
    .catch((err) => {
      dispatch(showNotification(err.response.data.message));
    });

export const updateRecipe = (recipe, index) => (dispatch) => {
  const {
    id,
    name,
    description,
    img_url,
    ingredients,
    instructions
  } = recipe;
  const update = {
    name,
    description,
    img_url,
    ingredients,
    instructions
  };

  dispatch(setFetching());
  return instance.put(`/recipes/${id}`, { update })
    .then((res) => {
      dispatch(batchActions([
        recieveUpdatedRecipe(res.data.recipe, index),
        unsetFetching()
      ]));
    })
    .catch((err) => {
      console.log(err.message || err.response.data.message);
      dispatch(unsetFetching());
    });
};
