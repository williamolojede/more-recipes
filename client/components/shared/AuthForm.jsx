import React from 'react';
import PropTypes from 'prop-types';

/**
 * @class AuthForm
 * @extends {Component}
 */
class AuthForm extends React.Component {
  /**
   * @param {node} e - element
   * @return {view} - renders a form
   */
  onFormSubmit(e) {
    e.preventDefault();

    const email = this.emailInput.value;
    const password = this.passwordInput.value;
    if (this.props.type === 'login') {
      this.props.authFormSubmit({ email, password });
    }

    this.authForm.reset();
  }

  /**
   * @return {view} - renders a form
   */
  render() {
    return (
      <form
        className="auth-form form"
        ref={(node) => { this.authForm = node; }}
        onSubmit={e => this.onFormSubmit(e)}
      >
        {
          this.props.type === 'signup' ?
            <input
              type="text"
              name="usr-name"
              className="form__input"
              placeholder="name"
              required
            />
            :
            null
        }
        <input
          id="usr-email"
          type="email"
          name="usr-email"
          className="form__input"
          placeholder="email"
          required
          ref={(node) => { this.emailInput = node; }}
        />
        <input
          id="usr-pswd"
          type="password"
          name="usr-pswd"
          className="form__input"
          placeholder="password"
          required
          ref={(node) => { this.passwordInput = node; }}
        />

        <button type="submit" className="form__submit z-depth-1">
          {
            this.props.type === 'signup' ?
              'Register'
              :
              'Login'
          }
        </button>
      </form>
    );
  }
}

AuthForm.propTypes = {
  type: PropTypes.string.isRequired,
  authFormSubmit: PropTypes.func.isRequired,
};

export default AuthForm;
