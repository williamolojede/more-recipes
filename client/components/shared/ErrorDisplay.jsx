import React from 'react';
import { errorDisplayPropTypes } from '../../config/proptypes';

const ErrorDisplay = props => (
  <div className="error z-depth-1">
    <span className="error__cancel">X</span>
    <p className="error__message">{ props.message }</p>
  </div>
);

ErrorDisplay.propTypes = errorDisplayPropTypes;

export default ErrorDisplay;
