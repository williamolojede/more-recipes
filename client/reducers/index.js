/**
 * Root Reducer
 */

import { combineReducers } from 'redux';
import { routerReducer } from 'react-router-redux';

// REDUCERS
import auth from './auth';
import isFetching from './isFetching';
import recipes from './recipes';

const rootReducer = combineReducers({
  auth,
  isFetching,
  recipes,
  router: routerReducer
});


export default rootReducer;
