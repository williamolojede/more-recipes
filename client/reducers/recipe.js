/**
 * Reducer for a Single Recipe
 */
import {
  RECEIVE_SINGLE_RECIPE,
  FETCH_SINGLE_RECIPE_ERROR,
  RESET_SINGLE_RECIPE
} from '../actions/types';
import defaultConfig from '../config/default';

export const initialState = {
  recipe: {
    id: 1,
    name: '',
    description: '',
    img_url: defaultConfig.recipeImgUrl,
    ingredients: [],
    instructions: [],
    upVoteCount: 0,
    downVoteCount: 0,
    favoriteCount: 0,
    viewCount: 0
  },
  errorMessage: ''
};

const singleRecipe = (state = initialState, { type, recipe, errorMessage }) => {
  switch (type) {
    case RECEIVE_SINGLE_RECIPE:
      return { ...state, ...{ recipe } };
    case FETCH_SINGLE_RECIPE_ERROR:
      return { ...state, ...{ errorMessage } };
    case RESET_SINGLE_RECIPE:
      return initialState;
    default:
      return state;
  }
};

export default singleRecipe;
