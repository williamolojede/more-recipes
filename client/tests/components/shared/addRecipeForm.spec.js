import React from 'react';
import { PureAddRecipeForm } from '../../../components/shared/AddRecipeForm';
import { initialState } from '../../../reducers/recipe';
import recipesMock from '../../__mocks__/recipes.mock';

jest.useFakeTimers();

function setup() {
  const props = {
    dispatch: jest.fn().mockReturnValue(Promise.resolve()),
    recipe: recipesMock.recipes[1],
    isEditMode: false,
    history: {
      push: jest.fn()
    }
  };
  const editModeProps = {
    dispatch: jest.fn().mockReturnValue(Promise.resolve()),
    recipe: recipesMock.recipes[1],
    isEditMode: true,
    match: {
      params: {
        recipeId: 5
      }
    },
    history: {
      push: jest.fn()
    }
  };

  const wrapper = mount(<PureAddRecipeForm {...props} />);
  const editModeWrapper = shallow(<PureAddRecipeForm {...editModeProps} />);

  return {
    wrapper,
    editModeWrapper,
    props,
    editModeProps
  };
}

describe('AddRecipeForm', () => {
  it('should render with right amount of elements', () => {
    const { wrapper } = setup();
    expect(wrapper).toMatchSnapshot();
  });

  it('should dispatch an action to reset the recipe stored in the store', () => {
    const { wrapper } = setup();
    const dispatchSpy = jest.spyOn(wrapper.instance().props, 'dispatch');
    wrapper.instance().componentWillUnmount();
    expect(dispatchSpy).toHaveBeenCalledTimes(1);
  });

  it('should remove item(ingredient or instruction) from state if delete button is clicked', () => {
    const { wrapper, props: { recipe } } = setup();
    const removeIngredientButtons = wrapper.find('.remove-item.remove-ingredients');
    const removeInstructionButtons = wrapper.find('.remove-item.remove-instructions');
    const removeItemSpy = jest.spyOn(wrapper.instance(), 'removeItem');

    expect(removeIngredientButtons.length).toBe(recipe.ingredients.length);
    expect(removeInstructionButtons.length).toBe(recipe.instructions.length);

    removeIngredientButtons.at(1).simulate('click', eventObject);
    removeInstructionButtons.at(1).simulate('click', eventObject);
    expect(removeItemSpy).toHaveBeenCalledTimes(2);

    const { ingredients, instructions } = wrapper.instance().state.recipe;
    expect(ingredients.length).toBe(recipe.ingredients.length - 1);
    expect(instructions.length).toBe(recipe.instructions.length - 1);
  });

  describe('addRecipeProp', () => {
    it('should add recipe name to the state', () => {
      const { wrapper } = setup();
      const nameInput = wrapper.find('#recipe-name');
      const addRecipePropSpy = jest.spyOn(wrapper.instance(), 'addRecipeProp');
      expect(nameInput.length).toBe(1);
      nameInput.simulate('change');
      expect(addRecipePropSpy).toHaveBeenCalledTimes(1);
    });

    it('should add recipe description to the state', () => {
      const { wrapper } = setup();
      const descriptionInput = wrapper.find('#recipe-description');
      const addRecipePropSpy = jest.spyOn(wrapper.instance(), 'addRecipeProp');
      expect(descriptionInput.length).toBe(1);
      descriptionInput.simulate('change');
      expect(addRecipePropSpy).toHaveBeenCalledTimes(1);
    });
  });

  describe('showInput/hideInput/addItem', () => {
    it('should render input fields when the add button is clicked and hide them when out of focus', () => {
      const { wrapper, props: { recipe } } = setup();
      const addNewButtons = wrapper.find('.add-new');

      // actions performed
      const showInputSpy = jest.spyOn(wrapper.instance(), 'showInput');
      const hideInputSpy = jest.spyOn(wrapper.instance(), 'hideInput');
      const addItemSpy = jest.spyOn(wrapper.instance(), 'addItem');


      expect(addNewButtons.length).toBe(2);

      // show input
      addNewButtons.forEach((button) => {
        button.simulate('click', eventObject);
      });
      expect(wrapper.instance().state.renderInput.ingredient).toBe(true);
      expect(wrapper.instance().state.renderInput.instruction).toBe(true);
      expect(showInputSpy).toHaveBeenCalledTimes(2);

      const inputFields = wrapper.find('.ingredient-instrution-input');
      expect(inputFields.length).toBe(2);

      // set value of input field
      wrapper.instance().addInstructionInput.value = 'add instruction';
      wrapper.instance().addIngredientInput.value = 'add ingredient';

      // add item and then hide input
      inputFields.forEach((input) => {
        input.simulate('keyUp', {
          ...eventObject,
          keyCode: 13,
          value: 'some text'
        });
        input.simulate('blur', eventObject);
      });

      expect(addItemSpy).toHaveBeenCalledTimes(2);
      expect(hideInputSpy).toHaveBeenCalledTimes(2);

      expect(wrapper.instance().state.renderInput.ingredient).toBe(false);
      expect(wrapper.instance().state.renderInput.instruction).toBe(false);

      const { ingredients, instructions } = wrapper.instance().state.recipe;
      expect(ingredients.length).toBe(recipe.ingredients.length + 1);
      expect(instructions.length).toBe(recipe.instructions.length + 1);
    });
  });

  describe('Edit Mode', () => {
    it('should render with right amount of elements', () => {
      const { editModeWrapper: wrapper } = setup();
      expect(wrapper).toMatchSnapshot();
    });

    it('should show a preloader while fetching recipe in edit mode', () => {
      const { editModeWrapper: wrapper } = setup();
      wrapper.setProps({ isFetching: true });
      expect(wrapper).toMatchSnapshot();
    });
  });

  describe('submitForm', () => {
    it('should call the update function when the form is submitted', () => {
      const { editModeWrapper: wrapper } = setup();
      // const AddRecipeForm = wrapper.find('AddRecipeForm');
      const form = wrapper.find('form');
      expect(form.length).toBe(1);
      const updateRecipeSpy = jest.spyOn(wrapper.instance(), 'updateRecipe');
      // const pushSpy = jest.spyOn(wrapper.instance().props.history, 'push');

      form.simulate('submit', eventObject);
      expect(updateRecipeSpy).toHaveBeenCalledTimes(1);
      jest.runAllTimers();
      // expect(pushSpy).toHaveBeenCalledTimes(1);
    });
    describe('AddRecipeMode', () => {
      it('should warn user when name or description field has not been filled', () => {
        const { wrapper } = setup();
        const form = wrapper.find('form');
        expect(form.length).toBe(1);

        const addRecipeSpy = jest.spyOn(wrapper.instance(), 'addRecipe');

        wrapper.setState({ recipe: initialState.recipe });
        form.simulate('submit', eventObject);
        expect(addRecipeSpy).toHaveBeenCalledTimes(1);
        expect(wrapper.instance().state.addRecipe.allow).toBe(false);
        expect(wrapper.instance().state.addRecipe.errorMessage).toBe('recipe name and description are required');
      });

      it('should warn user when ingredients or instructions has not been filled', () => {
        const { wrapper } = setup();
        const form = wrapper.find('form');
        expect(form.length).toBe(1);

        const addRecipeSpy = jest.spyOn(wrapper.instance(), 'addRecipe');

        wrapper.setState({ recipe: { ...recipesMock.recipes[1], instructions: [] } });
        form.simulate('submit', eventObject);
        expect(addRecipeSpy).toHaveBeenCalledTimes(1);
        expect(wrapper.instance().state.addRecipe.allow).toBe(false);
        expect(wrapper.instance().state.addRecipe.errorMessage).toBe('ingredients and instructions cannot be empty');
      });

      it('should submit the form if there are no errors', () => {
        const { wrapper } = setup();

        const form = wrapper.find('form');
        expect(form.length).toBe(1);

        const addRecipeSpy = jest.spyOn(wrapper.instance(), 'addRecipe');
        // const pushSpy = jest.spyOn(wrapper.instance().props.history, 'push');

        form.simulate('submit', eventObject);
        expect(addRecipeSpy).toHaveBeenCalledTimes(1);
        jest.runAllTimers();
        // expect(pushSpy).toHaveBeenCalledTimes(1);
      });
    });
  });
});
