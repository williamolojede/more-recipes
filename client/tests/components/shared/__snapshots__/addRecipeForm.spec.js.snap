// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AddRecipeForm Edit Mode should render with right amount of elements 1`] = `
<div
  className="container page-manage__add z-depth-1"
>
  <form
    className="my-recipe__add--form"
    onSubmit={[Function]}
  >
    <label
      className="my-recipe__add--form___name"
      htmlFor="recipe-name"
    >
      Name
      <input
        defaultValue="Chef John's Spaghetti with Red Clam Sauce dfbefadb nifek fedckltefdkletfdmlefdleff eifdanlkref"
        id="recipe-name"
        name="recipe-name"
        onChange={[Function]}
        placeholder="Chef John's Spaghetti with Red Clam Sauce dfbefadb nifek fedckltefdkletfdmlefdleff eifdanlkref"
        type="text"
      />
    </label>
    <div
      className="description-image"
    >
      <label
        className="my-recipe__add--form___desc"
        htmlFor="recipe-description"
      >
        Description
        <textarea
          defaultValue="This is one of my 'go to' recipes and it takes only about 15 minutes from start to finish. The only prep work really is mashing the garlic. The sauce is great on any pasta, but I prefer angel hair or spaghetti."
          id="recipe-description"
          onChange={[Function]}
          placeholder="e.g: short description..."
        />
      </label>
      <div
        className="my-recipe__add--form___img"
      >
        <p
          className="label"
        >
          Image
        </p>
        <ImageUpload
          altText="Chef John's Spaghetti with Red Clam Sauce dfbefadb nifek fedckltefdkletfdmlefdleff eifdanlkref"
          className=""
          defaultImgUrl="https://x.kinja-static.com/assets/images/logos/placeholders/default.png"
          imageUploadHandler={[Function]}
          imageUploadProgress="pending"
          imgUrl="http://images.media-allrecipes.com/userphotos/720x405/4478567.jpg"
        />
      </div>
    </div>
    <div
      className="ingredients-instruction"
    >
      <fieldset
        className="ingredients-form my-recipe__add--form___ingredients"
      >
        <legend>
          Ingredient(s):
        </legend>
        <ol>
          <li
            key="0"
          >
            2 tablespoons olive oil
            <a
              className="remove-item remove-ingredients"
              href="/delete"
              onClick={[Function]}
            >
              <i
                className="mdi mdi-delete"
              />
            </a>
          </li>
          <li
            key="1"
          >
            3 cloves garlic
            <a
              className="remove-item remove-ingredients"
              href="/delete"
              onClick={[Function]}
            >
              <i
                className="mdi mdi-delete"
              />
            </a>
          </li>
          <li
            key="2"
          >
            2 tablespoons capers
            <a
              className="remove-item remove-ingredients"
              href="/delete"
              onClick={[Function]}
            >
              <i
                className="mdi mdi-delete"
              />
            </a>
          </li>
          <li
            key="3"
          >
            1 tablespoon anchovy paste
            <a
              className="remove-item remove-ingredients"
              href="/delete"
              onClick={[Function]}
            >
              <i
                className="mdi mdi-delete"
              />
            </a>
          </li>
          <li
            key="4"
          >
            1 teaspoon red pepper flakes
            <a
              className="remove-item remove-ingredients"
              href="/delete"
              onClick={[Function]}
            >
              <i
                className="mdi mdi-delete"
              />
            </a>
          </li>
          <li
            key="5"
          >
            1 (24 ounce) jar tomato pasta sauce
            <a
              className="remove-item remove-ingredients"
              href="/delete"
              onClick={[Function]}
            >
              <i
                className="mdi mdi-delete"
              />
            </a>
          </li>
          <li
            key="6"
          >
            1/4 cup water
            <a
              className="remove-item remove-ingredients"
              href="/delete"
              onClick={[Function]}
            >
              <i
                className="mdi mdi-delete"
              />
            </a>
          </li>
          <li
            key="7"
          >
            1 cup good-quality crisp white wine
            <a
              className="remove-item remove-ingredients"
              href="/delete"
              onClick={[Function]}
            >
              <i
                className="mdi mdi-delete"
              />
            </a>
          </li>
          <li
            key="8"
          >
            2 (6.5 ounce) cans chopped clams, drained with juice reserved
            <a
              className="remove-item remove-ingredients"
              href="/delete"
              onClick={[Function]}
            >
              <i
                className="mdi mdi-delete"
              />
            </a>
          </li>
          <li
            key="9"
          >
            salt and pepper to taste
            <a
              className="remove-item remove-ingredients"
              href="/delete"
              onClick={[Function]}
            >
              <i
                className="mdi mdi-delete"
              />
            </a>
          </li>
          <li
            key="10"
          >
            1 pound dry spaghetti
            <a
              className="remove-item remove-ingredients"
              href="/delete"
              onClick={[Function]}
            >
              <i
                className="mdi mdi-delete"
              />
            </a>
          </li>
          <li
            key="11"
          >
            1/4 cup freshly grated Parmesan cheese
            <a
              className="remove-item remove-ingredients"
              href="/delete"
              onClick={[Function]}
            >
              <i
                className="mdi mdi-delete"
              />
            </a>
          </li>
          <li
            key="12"
          >
            Chopped fresh basil leaves or parsley
            <a
              className="remove-item remove-ingredients"
              href="/delete"
              onClick={[Function]}
            >
              <i
                className="mdi mdi-delete"
              />
            </a>
          </li>
        </ol>
        <a
          className="add-new add-new__ingredient"
          href="/add-new"
          onClick={[Function]}
        >
          <i
            className="mdi mdi-plus"
          />
          <span>
            Add an ingredient
          </span>
        </a>
      </fieldset>
      <fieldset
        className="instructions-form my-recipe__add--form___instructions"
      >
        <legend>
          Instruction(s):
        </legend>
        <ol>
          <li
            key="0"
          >
            Pour olive oil into cold skillet. Add garlic, capers, anchovy paste, and red pepper flakes. Place over medium heat and cook and stir until the oil is infused with the flavors of the mixture, about 5 minutes, taking care not to brown the garlic. Add tomato
            <a
              className="remove-item remove-instructions"
              href="/delete"
              onClick={[Function]}
            >
              <i
                className="mdi mdi-delete"
              />
            </a>
          </li>
          <li
            key="1"
          >
            Bring a large pot of lightly salted water to a boil. Cook spaghetti in the boiling water, stirring occasionally until almost al dente, or about 1 minute less than directed on package. Drain. Return pasta to the pot off heat.
            <a
              className="remove-item remove-instructions"
              href="/delete"
              onClick={[Function]}
            >
              <i
                className="mdi mdi-delete"
              />
            </a>
          </li>
          <li
            key="2"
          >
            Stir clams into the sauce. Pour sauce over pasta. Stir; cover the pot to allow it to finish cooking off heat and to allow the pasta will absorb some of the sauce, about 3 minutes. Stir in Parmesan cheese. Top with chopped fresh basil and/or parsley.
            <a
              className="remove-item remove-instructions"
              href="/delete"
              onClick={[Function]}
            >
              <i
                className="mdi mdi-delete"
              />
            </a>
          </li>
        </ol>
        <a
          className="add-new add-new__instruction"
          href="/add-new"
          onClick={[Function]}
        >
          <i
            className="mdi mdi-plus"
          />
          <span>
            Add an instruction
          </span>
        </a>
      </fieldset>
    </div>
    <div
      className="form-btns"
    >
      <SubmitButton
        handleClick={[Function]}
        submitLoading={false}
        submitText="Update"
      />
      <Link
        replace={false}
        to="/user/recipes"
      >
        or go back to your recipe list
      </Link>
    </div>
  </form>
</div>
`;

exports[`AddRecipeForm Edit Mode should show a preloader while fetching recipe in edit mode 1`] = `
<div
  className="container page-manage__add z-depth-1"
>
  <Preloader />
</div>
`;

exports[`AddRecipeForm should render with right amount of elements 1`] = `
<AddRecipeForm
  dispatch={[Function]}
  history={
    Object {
      "push": [Function],
    }
  }
  isEditMode={false}
  recipe={
    Object {
      "User": Object {
        "fullname": "Email Mail",
        "id": 11,
        "username": null,
      },
      "createdAt": "2017-11-09T09:08:15.109Z",
      "description": "This is one of my 'go to' recipes and it takes only about 15 minutes from start to finish. The only prep work really is mashing the garlic. The sauce is great on any pasta, but I prefer angel hair or spaghetti.",
      "downVoteCount": 1,
      "favoriteCount": 1,
      "id": 5,
      "img_url": "http://images.media-allrecipes.com/userphotos/720x405/4478567.jpg",
      "ingredients": Array [
        "2 tablespoons olive oil",
        "3 cloves garlic",
        "2 tablespoons capers",
        "1 tablespoon anchovy paste",
        "1 teaspoon red pepper flakes",
        "1 (24 ounce) jar tomato pasta sauce",
        "1/4 cup water",
        "1 cup good-quality crisp white wine",
        "2 (6.5 ounce) cans chopped clams, drained with juice reserved",
        "salt and pepper to taste",
        "1 pound dry spaghetti",
        "1/4 cup freshly grated Parmesan cheese",
        "Chopped fresh basil leaves or parsley",
      ],
      "instructions": Array [
        "Pour olive oil into cold skillet. Add garlic, capers, anchovy paste, and red pepper flakes. Place over medium heat and cook and stir until the oil is infused with the flavors of the mixture, about 5 minutes, taking care not to brown the garlic. Add tomato",
        "Bring a large pot of lightly salted water to a boil. Cook spaghetti in the boiling water, stirring occasionally until almost al dente, or about 1 minute less than directed on package. Drain. Return pasta to the pot off heat.",
        "Stir clams into the sauce. Pour sauce over pasta. Stir; cover the pot to allow it to finish cooking off heat and to allow the pasta will absorb some of the sauce, about 3 minutes. Stir in Parmesan cheese. Top with chopped fresh basil and/or parsley.",
      ],
      "name": "Chef John's Spaghetti with Red Clam Sauce dfbefadb nifek fedckltefdkletfdmlefdleff eifdanlkref",
      "owner": 11,
      "upVoteCount": 1,
      "updatedAt": "2018-01-10T10:54:26.519Z",
      "viewCount": 2,
    }
  }
>
  <div
    className="container page-manage__add z-depth-1"
  >
    <form
      className="my-recipe__add--form"
      onSubmit={[Function]}
    >
      <label
        className="my-recipe__add--form___name"
        htmlFor="recipe-name"
      >
        Name
        <input
          defaultValue="Chef John's Spaghetti with Red Clam Sauce dfbefadb nifek fedckltefdkletfdmlefdleff eifdanlkref"
          id="recipe-name"
          name="recipe-name"
          onChange={[Function]}
          placeholder="Chef John's Spaghetti with Red Clam Sauce dfbefadb nifek fedckltefdkletfdmlefdleff eifdanlkref"
          type="text"
        />
      </label>
      <div
        className="description-image"
      >
        <label
          className="my-recipe__add--form___desc"
          htmlFor="recipe-description"
        >
          Description
          <textarea
            defaultValue="This is one of my 'go to' recipes and it takes only about 15 minutes from start to finish. The only prep work really is mashing the garlic. The sauce is great on any pasta, but I prefer angel hair or spaghetti."
            id="recipe-description"
            onChange={[Function]}
            placeholder="e.g: short description..."
          />
        </label>
        <div
          className="my-recipe__add--form___img"
        >
          <p
            className="label"
          >
            Image
          </p>
          <ImageUpload
            altText="Chef John's Spaghetti with Red Clam Sauce dfbefadb nifek fedckltefdkletfdmlefdleff eifdanlkref"
            className=""
            defaultImgUrl="https://x.kinja-static.com/assets/images/logos/placeholders/default.png"
            imageUploadHandler={[Function]}
            imageUploadProgress="pending"
            imgUrl="http://images.media-allrecipes.com/userphotos/720x405/4478567.jpg"
          >
            <div
              className="image-upload "
            >
              <div
                className="image-upload__preview"
              >
                <img
                  alt="Chef John's Spaghetti with Red Clam Sauce dfbefadb nifek fedckltefdkletfdmlefdleff eifdanlkref"
                  className="image-upload__preview___image"
                  src="http://images.media-allrecipes.com/userphotos/720x405/4478567.jpg"
                />
                <div
                  className="image-upload__progress"
                />
              </div>
              <label
                className="z-depth-1"
                htmlFor="recipe-image"
              >
                Upload new picture
              </label>
              <input
                accept="image/gif, image/jpeg, image/png"
                id="recipe-image"
                name="filename"
                onChange={[Function]}
                type="file"
              />
            </div>
          </ImageUpload>
        </div>
      </div>
      <div
        className="ingredients-instruction"
      >
        <fieldset
          className="ingredients-form my-recipe__add--form___ingredients"
        >
          <legend>
            Ingredient(s):
          </legend>
          <ol>
            <li
              key="0"
            >
              2 tablespoons olive oil
              <a
                className="remove-item remove-ingredients"
                href="/delete"
                onClick={[Function]}
              >
                <i
                  className="mdi mdi-delete"
                />
              </a>
            </li>
            <li
              key="1"
            >
              3 cloves garlic
              <a
                className="remove-item remove-ingredients"
                href="/delete"
                onClick={[Function]}
              >
                <i
                  className="mdi mdi-delete"
                />
              </a>
            </li>
            <li
              key="2"
            >
              2 tablespoons capers
              <a
                className="remove-item remove-ingredients"
                href="/delete"
                onClick={[Function]}
              >
                <i
                  className="mdi mdi-delete"
                />
              </a>
            </li>
            <li
              key="3"
            >
              1 tablespoon anchovy paste
              <a
                className="remove-item remove-ingredients"
                href="/delete"
                onClick={[Function]}
              >
                <i
                  className="mdi mdi-delete"
                />
              </a>
            </li>
            <li
              key="4"
            >
              1 teaspoon red pepper flakes
              <a
                className="remove-item remove-ingredients"
                href="/delete"
                onClick={[Function]}
              >
                <i
                  className="mdi mdi-delete"
                />
              </a>
            </li>
            <li
              key="5"
            >
              1 (24 ounce) jar tomato pasta sauce
              <a
                className="remove-item remove-ingredients"
                href="/delete"
                onClick={[Function]}
              >
                <i
                  className="mdi mdi-delete"
                />
              </a>
            </li>
            <li
              key="6"
            >
              1/4 cup water
              <a
                className="remove-item remove-ingredients"
                href="/delete"
                onClick={[Function]}
              >
                <i
                  className="mdi mdi-delete"
                />
              </a>
            </li>
            <li
              key="7"
            >
              1 cup good-quality crisp white wine
              <a
                className="remove-item remove-ingredients"
                href="/delete"
                onClick={[Function]}
              >
                <i
                  className="mdi mdi-delete"
                />
              </a>
            </li>
            <li
              key="8"
            >
              2 (6.5 ounce) cans chopped clams, drained with juice reserved
              <a
                className="remove-item remove-ingredients"
                href="/delete"
                onClick={[Function]}
              >
                <i
                  className="mdi mdi-delete"
                />
              </a>
            </li>
            <li
              key="9"
            >
              salt and pepper to taste
              <a
                className="remove-item remove-ingredients"
                href="/delete"
                onClick={[Function]}
              >
                <i
                  className="mdi mdi-delete"
                />
              </a>
            </li>
            <li
              key="10"
            >
              1 pound dry spaghetti
              <a
                className="remove-item remove-ingredients"
                href="/delete"
                onClick={[Function]}
              >
                <i
                  className="mdi mdi-delete"
                />
              </a>
            </li>
            <li
              key="11"
            >
              1/4 cup freshly grated Parmesan cheese
              <a
                className="remove-item remove-ingredients"
                href="/delete"
                onClick={[Function]}
              >
                <i
                  className="mdi mdi-delete"
                />
              </a>
            </li>
            <li
              key="12"
            >
              Chopped fresh basil leaves or parsley
              <a
                className="remove-item remove-ingredients"
                href="/delete"
                onClick={[Function]}
              >
                <i
                  className="mdi mdi-delete"
                />
              </a>
            </li>
          </ol>
          <a
            className="add-new add-new__ingredient"
            href="/add-new"
            onClick={[Function]}
          >
            <i
              className="mdi mdi-plus"
            />
            <span>
              Add an ingredient
            </span>
          </a>
        </fieldset>
        <fieldset
          className="instructions-form my-recipe__add--form___instructions"
        >
          <legend>
            Instruction(s):
          </legend>
          <ol>
            <li
              key="0"
            >
              Pour olive oil into cold skillet. Add garlic, capers, anchovy paste, and red pepper flakes. Place over medium heat and cook and stir until the oil is infused with the flavors of the mixture, about 5 minutes, taking care not to brown the garlic. Add tomato
              <a
                className="remove-item remove-instructions"
                href="/delete"
                onClick={[Function]}
              >
                <i
                  className="mdi mdi-delete"
                />
              </a>
            </li>
            <li
              key="1"
            >
              Bring a large pot of lightly salted water to a boil. Cook spaghetti in the boiling water, stirring occasionally until almost al dente, or about 1 minute less than directed on package. Drain. Return pasta to the pot off heat.
              <a
                className="remove-item remove-instructions"
                href="/delete"
                onClick={[Function]}
              >
                <i
                  className="mdi mdi-delete"
                />
              </a>
            </li>
            <li
              key="2"
            >
              Stir clams into the sauce. Pour sauce over pasta. Stir; cover the pot to allow it to finish cooking off heat and to allow the pasta will absorb some of the sauce, about 3 minutes. Stir in Parmesan cheese. Top with chopped fresh basil and/or parsley.
              <a
                className="remove-item remove-instructions"
                href="/delete"
                onClick={[Function]}
              >
                <i
                  className="mdi mdi-delete"
                />
              </a>
            </li>
          </ol>
          <a
            className="add-new add-new__instruction"
            href="/add-new"
            onClick={[Function]}
          >
            <i
              className="mdi mdi-plus"
            />
            <span>
              Add an instruction
            </span>
          </a>
        </fieldset>
      </div>
      <div
        className="form-btns"
      >
        <SubmitButton
          handleClick={[Function]}
          submitLoading={false}
          submitText="Submit"
        >
          <button
            className="
    submit-btn z-depth-1
    
    "
            disabled={false}
            onClick={[Function]}
            type="submit"
          >
            <span>
              Submit
            </span>
            <Loading
              className="loading-icon"
              color="#fff"
              delay={1000}
              height="25px"
              type="spokes"
              width="25px"
            >
              <div
                className="loading-icon"
                dangerouslySetInnerHTML={
                  Object {
                    "__html": "<svg class=\\"icon-blank\\" xmlns=\\"http://www.w3.org/2000/svg\\" viewBox=\\"0 0 32 32\\"></svg>
",
                  }
                }
                style={
                  Object {
                    "fill": "#fff",
                    "height": "25px",
                    "width": "25px",
                  }
                }
              />
            </Loading>
          </button>
        </SubmitButton>
      </div>
    </form>
  </div>
</AddRecipeForm>
`;
